Feature/read only product variants

anitaa1990 opened the pull request:
Fixes #1540 by adding the read only product variant list to the product detail screen. 

### Notes:
- I followed the MVVM pattern.
- I've used the existing color palette in the app, rather than the ones specified in the design (since that will be handled separately).
- The product variant section will be displayed to all users once this PR is merged. But it is only clickable in debug builds. 
- Unit tests are added to `ProductVariantsViewModelTest`.

### Screenshots
<img width="300" src="https://user-images.githubusercontent.com/22608780/68450473-3bd90780-0211-11ea-8dd8-316689634659.png"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/68450481-3f6c8e80-0211-11ea-8a68-ba26cef56d68.png">

### Events
The following events have been added to the PR:
- `PRODUCT_DETAIL_VIEW_PRODUCT_VARIANTS_TAPPED` - User clicked on the variants section in the product detail screen.
- `PRODUCT_VARIANTS_LOADED` - The variants associated with a product was fetched successfully from the API.
- `PRODUCT_VARIANTS_LOAD_ERROR` - The request to fetch variants associated with a product from the API failed.
- `PRODUCT_VARIANTS_PULLED_TO_REFRESH` -  The user is viewing the product variants list and pulled to refresh.

These events have been added to the spreadsheet and will be validated/registered.


Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

nbradbury commented:
I get the following exception when I try to view a product's variants:

```
java.lang.IllegalStateException: Not a JSON Object: null
        at com.google.gson.JsonElement.getAsJsonObject(JsonElement.java:91)
        at org.wordpress.android.fluxc.network.rest.wpcom.wc.product.ProductRestClient.productVariationResponseToProductVariationModel(ProductRestClient.kt:499)
        at org.wordpress.android.fluxc.network.rest.wpcom.wc.product.ProductRestClient.access$productVariationResponseToProductVariationModel(ProductRestClient.kt:46)
        at org.wordpress.android.fluxc.network.rest.wpcom.wc.product.ProductRestClient$fetchProductVariations$request$1.invoke(ProductRestClient.kt:189)
        at org.wordpress.android.fluxc.network.rest.wpcom.wc.product.ProductRestClient$fetchProductVariations$request$1.invoke(ProductRestClient.kt:46)
        at org.wordpress.android.fluxc.network.rest.wpcom.jetpacktunnel.JetpackTunnelGsonRequest$buildGetRequest$wrappedListener$1.onResponse(JetpackTunnelGsonRequest.kt:118)
        at org.wordpress.android.fluxc.network.rest.wpcom.jetpacktunnel.JetpackTunnelGsonRequest$buildGetRequest$wrappedListener$1.onResponse(JetpackTunnelGsonRequest.kt:89)
        at org.wordpress.android.fluxc.network.rest.GsonRequest.deliverResponse(GsonRequest.java:53)
        at com.android.volley.ExecutorDelivery$ResponseDeliveryRunnable.run(ExecutorDelivery.java:102)
        at android.os.Handler.handleCallback(Handler.java:883)
        at android.os.Handler.dispatchMessage(Handler.java:100)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7356)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
```

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductVariantsAdapter.kt:
```diff
@@ -0,0 +1,89 @@
+    private val imageSize = context.resources.getDimensionPixelSize(R.dimen.product_icon_sz)
+    private val productVariantList = ArrayList<ProductVariant>()
+
+    override fun getItemId(position: Int) = productVariantList[position].remoteProductId
```
Since the IDs are unique, you could add this above:

```
init {
        setHasStableIds(true)
    }
```

anitaa1990 replied:
Missed this 🤦‍♀ Fixed in 5c6890a

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductVariantsAdapter.kt:
```diff
@@ -0,0 +1,89 @@
+
+        productVariant.imageUrl?.let {
+            val imageUrl = PhotonUtils.getPhotonImageUrl(it, imageSize, imageSize)
+            GlideApp.with(context)
```
np: you can create a single Glide instance and reuse it here. [Example](https://github.com/woocommerce/woocommerce-android/blob/5ed5c6777c4f520fd9d916b097c20e4d797837f7/WooCommerce/src/main/kotlin/com/woocommerce/android/widgets/WCProductImageGalleryView.kt#L90)

anitaa1990 replied:
The [example](https://github.com/woocommerce/woocommerce-android/blob/5ed5c6777c4f520fd9d916b097c20e4d797837f7/WooCommerce/src/main/kotlin/com/woocommerce/android/widgets/WCProductImageGalleryView.kt#L90) points to an implementation with `RecyclerView` 🤔 

I've used this [SO answer](https://stackoverflow.com/a/32887693/9862062) as a reference and added logic to pass the `GlideRequests` as a parameter to the adapter instead in f61fa1a

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductVariantsAdapter.kt:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    private class ProductVariantItemDiffUtil(
+        val items: List<ProductVariant>,
```
For clarity, how about naming these `oldList` and `newList`?

anitaa1990 replied:
👍 Done in 5c6890a

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductVariantsAdapter.kt:
```diff
@@ -0,0 +1,89 @@
+        override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
+            val oldItem = items[oldItemPosition]
+            val newItem = result[newItemPosition]
+            return oldItem.stockQuantity == newItem.stockQuantity &&
```
Should the prices be compared here as well?

anitaa1990 replied:
Good point! Fixed in 1d118aa

nbradbury commented:
Really nice work on this, @anitaa1990! I'm done with my first round - just a few minor issues plus a FluxC crash.

anitaa1990 commented:
Thanks for the quick review @nbradbury! I've addressed all the issues and opened a [PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1423) for the FluxC crash. Once that PR is merged, I will update the hash to this PR 👍 

Ready for another round

anitaa1990 commented:
> Sorry I missed this the first time, but I think we also want the first comparison to check the remoteVariationId.

Can't believe I missed that 🤦‍♀ Fixed in f0148ca and updated the FluxC hash. Ready for another round 👍
